src/main/java/de/tum/cit/aet/artemis/programming/service/localvc/GitPublickeyAuthenticatorService.java

```diff
@@ -32,48 +35,91 @@ public class GitPublickeyAuthenticatorService implements PublickeyAuthenticator
 
     private final Optional<SharedQueueManagementService> localCIBuildJobQueueService;
 
-    public GitPublickeyAuthenticatorService(UserRepository userRepository, Optional<SharedQueueManagementService> localCIBuildJobQueueService) {
+    private final UserSshPublicKeyRepository userSshPublicKeyRepository;
+
+    public GitPublickeyAuthenticatorService(UserRepository userRepository, Optional<SharedQueueManagementService> localCIBuildJobQueueService,
+            UserSshPublicKeyRepository userSshPublicKeyRepository) {
         this.userRepository = userRepository;
         this.localCIBuildJobQueueService = localCIBuildJobQueueService;
+        this.userSshPublicKeyRepository = userSshPublicKeyRepository;
     }
 
     @Override
     public boolean authenticate(String username, PublicKey publicKey, ServerSession session) {
         String keyHash = HashUtils.getSha512Fingerprint(publicKey);
-        var user = userRepository.findBySshPublicKeyHash(keyHash);
-        if (user.isPresent()) {
-            try {
-                // Retrieve the stored public key string
-                String storedPublicKeyString = user.get().getSshPublicKey();
-
-                // Parse the stored public key string
-                AuthorizedKeyEntry keyEntry = AuthorizedKeyEntry.parseAuthorizedKeyEntry(storedPublicKeyString);
-                PublicKey storedPublicKey = keyEntry.resolvePublicKey(null, null, null);
-
-                // Compare the stored public key with the provided public key
-                if (Objects.equals(storedPublicKey, publicKey)) {
-                    log.debug("Found user {} for public key authentication", user.get().getLogin());
-                    session.setAttribute(SshConstants.USER_KEY, user.get());
-                    session.setAttribute(SshConstants.IS_BUILD_AGENT_KEY, false);
-                    return true;
-                }
-                else {
-                    log.warn("Public key mismatch for user {}", user.get().getLogin());
-                }
+        var userSshPublicKey = userSshPublicKeyRepository.findByKeyHash(keyHash);
+        return userSshPublicKey.map(sshPublicKey -> {
+            ZonedDateTime expiryDate = sshPublicKey.getExpiryDate();
+            if (expiryDate == null || expiryDate.isAfter(ZonedDateTime.now())) {
+                return authenticateUser(sshPublicKey, publicKey, session);
```

Comment:
I think we should clearly communicated expired keys to end users. Especially when students set up expiry dates and they co-incidence with an exam, this could be very hard to understand otherwise why the authentication does not work. I hope it's possible to clearly communicate the issue to the respective git client

Reply:
It would even make sense to implement a scheduled email notification shortly before the ssh key expires. A simple implementation could be a scheduled activity on the `scheduling` node (with a configurable cron, default could be once a day), which checks for expiring ssh keys within the next 7 days and notifies the respective users via email about the situation. This email should then contain a link to the ssh settings page

Reply:
Notice, that you might need to add an index to the expiry date column to quickly retrieve all keys that expire directly within the database

Reply:
This sounds like a very nice feature but might exceed the scope of this PR 👍

Reply:
Maybe you can even show in the overview page for the keys whether keys are (about to be) expired?

Reply:
![image](https://github.com/user-attachments/assets/e20c8e2b-81de-4fb2-a520-509cec681e50)
So in the overview you can see the expiry date of keys (if set) directly under the key. 

To also make it clear to the user that some keys are expired, I added a warning to the code button:

![image](https://github.com/user-attachments/assets/c9379689-609c-4cb7-b4db-d584216da33a)